glean-core/src/metrics/event.rs

```diff
@@ -0,0 +1,132 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::collections::HashMap;
+use std::iter::Iterator;
+
+use serde_json::{json, Value as JsonValue};
+
+use crate::error_recording::{record_error, ErrorType};
+use crate::event_database::RecordedEventData;
+use crate::metrics::MetricType;
+use crate::CommonMetricData;
+use crate::Glean;
+
+const MAX_LENGTH_EXTRA_KEY_VALUE: usize = 100;
+
+/// An event metric.
+///
+/// Events allow recording of e.g. individual occurences of user actions, say
+/// every time a view was open and from where. Each time you record an event, it
+/// records a timestamp, the event's name and a set of custom values.
+#[derive(Clone, Debug)]
+pub struct EventMetric {
+    meta: CommonMetricData,
+    allowed_extra_keys: Vec<String>,
+}
+
+impl MetricType for EventMetric {
+    fn meta(&self) -> &CommonMetricData {
+        &self.meta
+    }
+
+    fn meta_mut(&mut self) -> &mut CommonMetricData {
+        &mut self.meta
+    }
+}
+
+impl EventMetric {
+    /// Create a new event metric.
+    pub fn new(meta: CommonMetricData, allowed_extra_keys: Vec<String>) -> Self {
+        Self {
+            meta,
+            allowed_extra_keys,
+        }
+    }
+
+    /// Record an event.
+    ///
+    /// ## Arguments
+    ///
+    /// * `glean`:  The Glean instance this metric belongs to.
+    /// * `timestamp`: A monotonically increasing timestamp, in nanoseconds.
+    ///   This must be provided since the actual recording of the event may
+    ///   happen some time later than the moment the event occurred.
+    /// * `extra`: A HashMap of (key, value) pairs. The key is an index
+    ///   into the metric's allowed_extra_keys vector where the key's string
+    ///   is looked up.
+    pub fn record(&self, glean: &Glean, timestamp: u64, extra: Option<HashMap<usize, String>>) {
+        if !self.should_record(glean) {
+            return;
+        }
+
+        let extra_strings = extra.and_then(|extra| {
+            Some(
+                extra
+                    .iter()
+                    .map(|(k, v)| (self.allowed_extra_keys.get(*k).unwrap(), v))
+                    .map(|(k, v)| (k.to_string(), self.truncate_value(glean, k, v).to_string()))
+                    .collect(),
+            )
+        });
+
+        glean
+            .event_storage()
+            .record(glean, &self.meta, timestamp, extra_strings);
+    }
+
+    fn truncate_value<'a>(&self, glean: &Glean, key: &str, value: &'a str) -> &'a str {
```

Comment:
nit: can you add a tiny comment describing what does this do? It currently does a bit more than just truncating (i.e. error reporting). Plus, it returns the truncated value.